components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,329 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/eth_allowance_manager.h"
+
+#include <algorithm>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "brave/components/brave_wallet/browser/blockchain_registry.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_service.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/json_rpc_service.h"
+#include "brave/components/brave_wallet/browser/keyring_service.h"
+#include "brave/components/brave_wallet/browser/pref_names.h"
+#include "brave/components/brave_wallet/common/hash_utils.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+
+namespace brave_wallet {
+
+EthAllowanceManager::EthAllowanceManager(BraveWalletService* wallet_service,
+                                         JsonRpcService* json_rpc_service,
+                                         KeyringService* keyring_service,
+                                         PrefService* prefs)
+    : wallet_service_(wallet_service),
+      json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs),
+      weak_ptr_factory_(this) {}
+
+EthAllowanceManager::~EthAllowanceManager() = default;
+
+void EthAllowanceManager::DiscoverEthAllowancesOnAllSupportedChains() {
+  if (is_eth_allowance_discovering_running_) {
+    return;
+  }
+  is_eth_allowance_discovering_running_ = true;
+
+  const auto keyring_info = keyring_service_->GetKeyringInfoSync(
+      brave_wallet::mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    return;
```

Comment:
We can't just return directly, we need to call CompleteDiscovedEthAllowance to reset the running state and fire completion event.
And if we change the API as I suggested in the mojom file, we should run the callback in all cases including error paths.